core/service.go

```diff
@@ -324,25 +334,55 @@ func (s *Service) handleReceivedMessage(msg network.Message) (err error) {
 // announce messages (block announce messages include the header but the full
 // block is required to execute `core_execute_block`).
 func (s *Service) ProcessBlockAnnounceMessage(msg network.Message) error {
-	data := msg.(*network.BlockAnnounceMessage)
+	blockAnnounceMessage, ok := msg.(*network.BlockAnnounceMessage)
+	if !ok {
+		return errors.New("could not cast network.Message to BlockAnnounceMessage")
+	}
 
-	header, err := types.NewHeader(data.ParentHash, data.Number, data.StateRoot, data.ExtrinsicsRoot, data.Digest)
-	if err != nil {
-		return err
+	if s.blockState == nil {
+		return errors.New("ProcessBlockAnnounceMessage error: blockState is nil")
 	}
 
-	// TODO: check if we should send block request message
-	_, err = s.blockState.GetHeader(header.Hash())
-	if err != nil && err.Error() == "Key not found" {
-		err = s.blockState.SetHeader(header)
+	latestBlockNum := s.blockState.LatestHeader().Number
+	messageBlockNumMinusOne := big.NewInt(0).Sub(blockAnnounceMessage.Number, big.NewInt(1))
+
+	// check if we should send block request message
+	if latestBlockNum.Cmp(messageBlockNumMinusOne) == -1 {
+
+		//generate random ID
+		s1 := rand.NewSource(uint64(time.Now().UnixNano()))
+		seed := rand.New(s1).Uint64()
+		randomID := mrand.New(mrand.NewSource(int64(seed))).Uint64()
+
+		currentHash, _ := common.HexToBytes(s.blockState.LatestHeader().Hash().String())
```

Comment:
the `.String()` and `HexToBytes` are redundant, .Hash() is already `[32]byte`, to get a `[32]byte` as a slice you can do `[:]` at the end

Reply:
eg `currentHash := s.blockState.LatestHeader().Hash()`

Reply:
yes, thanks.